File: IsPalindrome/IsPalindrome.py
import re

def isPalindromeFirst(string: str) -> bool: # best
    return re.sub(r'\W+', '', string).lower() == re.sub(r'\W+', '', string).lower()[::-1]

def isPalindromeSecond(string: str) -> bool:
    return list(re.sub(r'\W+', '', string).lower()) == list(reversed(list(re.sub(r'\W+', '', string).lower())))

def isPalindromeSecondAlt(string: str) -> bool:
    return "".join(list(re.sub(r'\W+', '', string).lower())) == "".join(list(reversed(list(re.sub(r'\W+', '', string).lower()))))

def isPalindromeThird(string: str) -> bool:
    reversed_list = list(re.sub(r'\W+', '', string).lower())
    reversed_list.reverse()
    return list(re.sub(r'\W+', '', string).lower()) == reversed_list

def isPalindromeThirdAlt(string: str) -> bool:
    reversed_list = list(re.sub(r'\W+', '', string).lower())
    reversed_list.reverse()
    return "".join(list(re.sub(r'\W+', '', string).lower())) == "".join(reversed_list)

def isPalindromeFourth(string: str) -> bool:
    clean_string = re.sub(r'\W+', '', string).lower()
    reversed_string = ''
    for i in range(len(clean_string), 0, -1):
        reversed_string += clean_string[i-1]
    return clean_string == reversed_string

if __name__ == '__main__':
    print(isPalindromeFirst("Red roses run no risk, sir, on Nurse's order"))
    print(isPalindromeSecond("Red roses run no risk, sir, on Nurse's order"))
    print(isPalindromeSecondAlt("Red roses run no risk, sir, on Nurse's order"))
    print(isPalindromeThird("Red roses run no risk, sir, on Nurse's order"))
    print(isPalindromeThirdAlt("Red roses run no risk, sir, on Nurse's order"))
    print(isPalindromeFourth("Red roses run no risk, sir, on Nurse's order"))

File: IsPalindrome/IsPalindrome.go
package main

import (
	"fmt"
	"regexp"
	"strings"
)

func IsPalindromeFirst(str string) bool {
	rePattern := regexp.MustCompile(`\W+`)
	cleanStr := strings.ToLower(rePattern.ReplaceAllString(str, ""))
	strRune := []rune(cleanStr)
	for i, j := 0, len(strRune)-1; i < j; i, j = i+1, j-1 {
		strRune[i], strRune[j] = strRune[j], strRune[i]
	}
	reverseStr := string(strRune)
	return cleanStr == reverseStr
}

func IsPalindromeSecond(str string) bool {
	rePattern := regexp.MustCompile(`\W+`)
	cleanStr := strings.ToLower(rePattern.ReplaceAllString(str, ""))
	reverseStr := ""
	for _, v := range cleanStr {
		reverseStr = string(v) + reverseStr
	}
	return cleanStr == reverseStr
}

func IsPalindromeThird(str string) bool {
	rePattern := regexp.MustCompile(`\W+`)
	cleanStr := strings.ToLower(rePattern.ReplaceAllString(str, ""))
	reverseStr := []byte{}
	for i := len(cleanStr) - 1; i >= 0; i-- {
		reverseStr = append(reverseStr, cleanStr[i])
	}
	return cleanStr == string(reverseStr)
}

func main() {
	fmt.Println(IsPalindromeFirst("Red roses run no risk, sir, on Nurse's order"))
	fmt.Println(IsPalindromeSecond("Red roses run no risk, sir, on Nurse's order"))
	fmt.Println(IsPalindromeThird("Red roses run no risk, sir, on Nurse's order"))
}
